pkg/cache/sql/encryption/encrypt.go

```diff
@@ -133,50 +121,56 @@ func createGCMCypher(key []byte) (cipher.AEAD, error) {
 	return aead, nil
 }
 
-func (m *Manager) isWriteCounterOverMax() bool {
-	return m.writesCounter > maxWriteCount
-}
-
-// fetchUpToDateDataKey returns the current data key if writeCounter has not exceeded maxWriteCount. If it has then the
-// data key is rotated. Before exiting the writesCounter is incremented by 1. This means that one use of the key has
-// been accounted for. This is done so encryption does not block other processes from encrypting. The consequence is a
-// key should only be used once and fetchUpToDateDataKey should be used to ensure an up-to-date key for any subsequent
-// encryption.
-func (m *Manager) fetchUpToDateDataKey() ([]byte, error) {
+// fetchActiveDataKey returns the current data key and its key ID.
+// Each call results in activeKeyCounter being incremented by 1. When the
+// the activeKeyCounter exceeds maxWriteCount, the active data key is
+// rotated - before being returned.
+func (m *Manager) fetchActiveDataKey() ([]byte, uuid.UUID, error) {
 	m.lock.Lock()
-	defer m.lock.Unlock()
```

Comment:
What pros and cons do you see with using `defer` here?

Reply:
The switch was mostly due to the unnecessary performance impact of deferring. But happy to switch back if maintainability is more important in this case.